Start each request on fresh ExecutionContext

benaadams opened the pull request:
Check with Hide Whitespace on; as indents changed.

Fixes #13991

Per request allocations for Plaintext sample remains the same

Allocations for 17k requests across 256 connections

![image](https://user-images.githubusercontent.com/1142958/65285046-43463380-db33-11e9-89fa-dae5c433ae33.png)

benaadams committed:
Start each request on fresh ExecutionContext

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -608,103 +615,124 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
-                    // Run the application code for this request
-                    await application.ProcessRequestAsync(context);
+                // Clear the TContext for the request
+                application.DisposeContext(requestData.Context, _applicationException);
```
Given that it's pluggable, should CreateContext and DisposeContext be called with the `cleanExecutionContext`?

benaadams replied:
I suppose

benaadams replied:
Done

benaadams commented:
Per request allocations for Plaintext sample remains similar
![image](https://user-images.githubusercontent.com/1142958/65285046-43463380-db33-11e9-89fa-dae5c433ae33.png)

benaadams commented:
Cleaned up the encapsulation

benaadams commented:
Since we are calling an extra async method anyway; realised it could be greatly simplified.

(ignore whitespace)

benaadams commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -608,103 +610,111 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
+        }
 
-                KestrelEventSource.Log.RequestStop(this);
+        private async ValueTask ProcessRequest<TContext>(IHttpApplication<TContext> application)
```
`Task` is a more performant return here than `ValueTask` and will allocate the same; i.e none in the sync completion and the async box in the async wait scenario (subject to https://github.com/dotnet/coreclr/pull/26310 when `ValueTask` becomes better as it doesn't allocate in async)

halter73 commented:
So much nicer! We probably still want to take the tests from #14027 though.

halter73 reviewed (approved):
One subtle difference between this and #14027 is that the Response.OnCompleted callback will run with the same ExecutionContext as the middleware if the middleware isn't "async". With #14027, the Response.OnCompleted gets a fresh ExecutionContext no matter what. I think this probably is fine though as long as we add a test and don't change behavior.

benaadams commented:
> One subtle difference between this and #14027 is that the Response.OnCompleted callback will run with the same ExecutionContext as the middleware if the middleware isn't "async". 

Might even be desirable as it allows you to set a asynclocal (or "request" local) reference type in `OnStarted`; have it flow through the request and then be able to check its properties in `OnCompleted`

benaadams commented:
Hmm... make that in the first middleware :)

benaadams commented:
Added tests

> I think this probably is fine though as long as we add a test and don't change behaviour.

Also added compat test and commented it as such https://github.com/aspnet/AspNetCore/pull/14146/files#diff-d47b9baa7ec5338570136a89e5e8a4deR441-R444

```csharp
public async Task ExecutionContextMutationsOfValueTypeFlowThroughNonAsyncEvents()
{
    // Note: This test is for future compat behviour checking
    // to ensure non-async OnStarting/OnCompleted flow AsyncLocal changes out
```

davidfowl commented:
We need to run performance tests

benaadams commented:
> We need to run performance tests

Don't think it will be great :(

However... might as well make the OnStarting/OnComplete async as they won't regress performance; since normally aren't called?

benaadams commented:
Changed it to flow normally for non-async events also; since they won't run if you haven't defined any so it shouldn't effect performance if you choose not to use them.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -733,108 +743,55 @@ public void OnCompleted(Func<object, Task> callback, object state)
+            if (onCompleted?.Count > 0)
             {
-                return Task.CompletedTask;
+                return ProcessEvents(this, onCompleted);
```
If one of the other onCompleted callbacks throws, do we still continue looping over the other callbacks and executing? Unlike the onStarting callbacks, it's important that every onCompleted callback is invoked. I thought we had a test for this.

halter73 replied:
Nvm. I just saw this was calling a local function.

anurse commented:
@halter73 are you good to merge this in for 5.0?

davidfowl commented:
> @halter73 are you good to merge this in for 5.0?

Not without performance tests.

benaadams commented:
Is that auto perf test ci thing working? https://github.com/aspnet/AspNetCore/pull/16999 (seems to be having versioning issues)

davidfowl commented:
Maybe we can use the in memory unit tests.

benaadams commented:
Lemme rebase

benaadams commented:
Changed to ValueTask since https://github.com/dotnet/coreclr/pull/26310 was merged; but would be good to also test the perf with the env variable `DOTNET_SYSTEM_THREADING_POOLASYNCVALUETASKS` set to 1 to see have that changes things; probably on something with some async (database calls?) and then feedback to https://github.com/dotnet/coreclr/issues/27403

anurse commented:
> Is that auto perf test ci thing working?

@halter73 says soon, so we'll try running it on this when it's working.

anurse commented:
@aspnet-hello benchmark

anurse commented:
@halter73 :(

halter73 commented:
@aspnet-hello benchmark

benaadams commented:
329,492 -> 354,384?

Even when I make it slower I make it faster? 😅😎

davidfowl commented:
Whattttt

davidfowl commented:
@aspnet-hello benchmark

benaadams commented:
@aspnet-hello benchmark

benaadams commented:
😢

benaadams commented:
Is there a way to test with the environment variable `DOTNET_SYSTEM_THREADING_POOLASYNCVALUETASKS` set to `1`?

https://github.com/dotnet/coreclr/issues/27403

anurse commented:
Seems like this is within the noise. Are we ok taking it and monitoring the rolling benchmarks? @halter73 @davidfowl .

anurse commented:
/azp run

Tratcher commented:
/azp run

benaadams commented:
@aspnet-hello benchmark

anurse commented:
@davidfowl are you good with the perf on this? I'd like to close it out and preferably by merging it ;P

anurse commented:
Ping @davidfowl @halter73

anurse commented:
Hearing nothing from @davidfowl, I'm going to go ahead and tell @halter73 he can merge away ;).

davidfowl commented:
Merge it we can measure things now

benaadams commented:
Rebased

benaadams commented:
@aspnet-hello benchmark

benaadams commented:
@aspnet-hello benchmark

BIOsborne commented:
I have observed an issue (on AspNetCore version 3.1.13) with AsyncLocal context being leaked between requests. (Never in parallel, the Context from one request which finishes appears to be present at the start of a subsequent one).
  
I have not attempted (yet) to reproduce this locally, but have observed it reliably in our Production systems.
Interestingly - we are running 50% production traffic through Kestrel (running on Linux) and 50% running through Azure Windows hosted with IIS (as part of a live migration process). The issue appears only with Kestrel on Linux. Not sure if that is relevant.
  
I have seen this issue is closed now: https://github.com/dotnet/aspnetcore/issues/13991
I assume we would not expect this issue to be live still?
If this is of interest, I can create a new issue, but won't do this if there are live issues already open around this that I have not been able to find.

A little more context - the AsyncLocal state tracked in our case gets initialised in Log sink which is NOT in an Async method. (The only reason I mention that is that some comments in the linked issue refer to the method setting the AsyncLocal state should be async.
(https://github.com/dotnet/aspnetcore/issues/13991#issuecomment-531633911)

Any input gratefully received - hope this isn't wasting your time!
Thanks.

alefranz commented:
@BIOsborne this fix is in .NET 5 and has not been backported to 3.1.
As far as I am aware, the only way to work around the problem in 3.1 is what David suggested in the comment you pointed out.

BIOsborne commented:
> @BIOsborne this fix is in .NET 5 and has not been backported to 3.1.
> As far as I am aware, the only way to work around the problem in 3.1 is what David suggested in the comment you pointed out.

Thanks @alefranz for quick reply - good to know.
We have been able to work around this in 3.1.* for now no problem.
